File: api/google.js
/**
 * Google Identity Services ID-token verify + local USER provisioning.
 * Set GOOGLE_CLIENT_ID (OAuth web client) to enable /api/auth/google.
 */
const crypto = require("crypto");

class GoogleAuthError extends Error {
  constructor(message, status = 400) {
    super(message);
    this.status = status;
  }
}

function googleClientId() {
  return String(process.env.GOOGLE_CLIENT_ID || "").trim();
}

let jwksCache = { keys: null, exp: 0 };

async function googleJwks() {
  if (jwksCache.keys && Date.now() < jwksCache.exp) return jwksCache.keys;
  const res = await fetch("https://www.googleapis.com/oauth2/v3/certs", {
    headers: { Accept: "application/json" },
    signal: AbortSignal.timeout(8000),
  });
  if (!res.ok) throw new GoogleAuthError("Google sign-in could not be verified.");
  const data = await res.json();
  const keys = data.keys || [];
  jwksCache = { keys, exp: Date.now() + 60 * 60 * 1000 };
  return keys;
}

function b64urlJson(part) {
  return JSON.parse(Buffer.from(part, "base64url").toString("utf8"));
}

async function verifyGoogleIdToken(credential) {
  const clientId = googleClientId();
  if (!clientId) throw new GoogleAuthError("Google sign-in is not configured.", 503);
  if (!credential || typeof credential !== "string") {
    throw new GoogleAuthError("Google credential is required.");
  }
  const parts = credential.split(".");
  if (parts.length !== 3) throw new GoogleAuthError("Google sign-in could not be verified.");

  const [h, p, sig] = parts;
  let header;
  let payload;
  try {
    header = b64urlJson(h);
    payload = b64urlJson(p);
  } catch {
    throw new GoogleAuthError("Google sign-in could not be verified.");
  }

  const keys = await googleJwks();
  const jwk = keys.find((k) => k.kid === header.kid);
  if (!jwk) throw new GoogleAuthError("Google sign-in could not be verified.");

  const key = crypto.createPublicKey({ key: jwk, format: "jwk" });
  const ok = crypto.verify("RSA-SHA256", Buffer.from(`${h}.${p}`), key, Buffer.from(sig, "base64url"));
  if (!ok) throw new GoogleAuthError("Google sign-in could not be verified.");

  const issOk =
    payload.iss === "accounts.google.com" || payload.iss === "https://accounts.google.com";
  const aud = payload.aud;
  const audOk = aud === clientId || (Array.isArray(aud) && aud.includes(clientId));
  const expOk = typeof payload.exp === "number" && payload.exp * 1000 > Date.now() - 10_000;
  if (!issOk || !audOk || !expOk) {
    throw new GoogleAuthError("Google sign-in could not be verified.");
  }
  if (!payload.sub || !payload.email || !payload.email_verified) {
    throw new GoogleAuthError("Google did not provide a verified email.");
  }
  return payload;
}

function uniqueUsername(db, email, name) {
  const local = String(email || "").split("@")[0];
  let base = String(local || name || "user")
    .replace(/[^A-Za-z0-9._]/g, "")
    .slice(0, 24);
  if (!base) base = "user";
  if (/^\d/.test(base)) base = `u${base}`.slice(0, 24);
  let candidate = base;
  let n = 2;
  while (db.prepare("SELECT id FROM users WHERE username = ? COLLATE NOCASE").get(candidate)) {
    const suffix = String(n);
    candidate = `${base.slice(0, Math.max(1, 24 - suffix.length))}${suffix}`;
    n += 1;
    if (n > 10_000) throw new GoogleAuthError("Could not allocate a username.");
  }
  return candidate;
}

function userFromGoogle(db, info) {
  const sub = String(info.sub);
  const email = String(info.email || "").trim().toLowerCase();

  let row = db.prepare("SELECT * FROM users WHERE google_sub = ?").get(sub);
  if (row) {
    if (email && !row.email) {
      db.prepare("UPDATE users SET email = ? WHERE id = ?").run(email, row.id);
      row = db.prepare("SELECT * FROM users WHERE id = ?").get(row.id);
    }
    return { user: row, created: false };
  }

  if (email) {
    const byEmail = db
      .prepare("SELECT * FROM users WHERE lower(email) = lower(?) AND email != ''")
      .get(email);
    if (byEmail) {
      if (byEmail.google_sub && byEmail.google_sub !== sub) {
        throw new GoogleAuthError("This Google account cannot be linked.");
      }
      db.prepare("UPDATE users SET google_sub = ? WHERE id = ?").run(sub, byEmail.id);
      return { user: db.prepare("SELECT * FROM users WHERE id = ?").get(byEmail.id), created: false };
    }
  }

  const username = uniqueUsername(db, email, info.name);
  const infoIns = db
    .prepare(
      `INSERT INTO users (username, email, password_hash, role, avatar_seed, google_sub)
       VALUES (?, ?, '', 'USER', ?, ?)`,
    )
    .run(username, email, username, sub);
  return {
    user: db.prepare("SELECT * FROM users WHERE id = ?").get(infoIns.lastInsertRowid),
    created: true,
  };
}

module.exports = {
  GoogleAuthError,
  googleClientId,
  verifyGoogleIdToken,
  userFromGoogle,
  uniqueUsername,
};

File: backend/apps/accounts/google.py
"""Google ID-token verification and local USER provisioning."""

from __future__ import annotations

import re

from django.conf import settings
from django.contrib.auth import get_user_model
from rest_framework.exceptions import APIException, ValidationError

from .activation import mark_email_verified
from .models import Role

User = get_user_model()


class GoogleAuthError(Exception):
    def __init__(self, detail: str, status: int = 400):
        super().__init__(detail)
        self.detail = detail
        self.status = status


def google_client_id() -> str:
    return (getattr(settings, "GOOGLE_CLIENT_ID", "") or "").strip()


def verify_google_id_token(credential: str) -> dict:
    """Verify a GIS ID token and return the payload (sub, email, ...)."""
    client_id = google_client_id()
    if not client_id:
        raise GoogleAuthError("Google sign-in is not configured.", status=503)
    if not credential or not isinstance(credential, str):
        raise GoogleAuthError("Google credential is required.")

    try:
        from google.auth.transport import requests as google_requests
        from google.oauth2 import id_token
    except ImportError as exc:  # pragma: no cover
        raise GoogleAuthError("Google sign-in is not available.", status=503) from exc

    try:
        info = id_token.verify_oauth2_token(
            credential,
            google_requests.Request(),
            client_id,
            clock_skew_in_seconds=10,
        )
    except Exception as exc:
        raise GoogleAuthError("Google sign-in could not be verified.") from exc

    iss = info.get("iss")
    if iss not in ("accounts.google.com", "https://accounts.google.com"):
        raise GoogleAuthError("Google sign-in could not be verified.")
    if not info.get("email") or not info.get("email_verified"):
        raise GoogleAuthError("Google did not provide a verified email.")
    if not info.get("sub"):
        raise GoogleAuthError("Google sign-in could not be verified.")
    return info


def _username_base(email: str, name: str | None) -> str:
    local = (email or "").split("@", 1)[0]
    raw = local or (name or "") or "user"
    cleaned = re.sub(r"[^A-Za-z0-9._]", "", raw)[:24]
    if not cleaned:
        cleaned = "user"
    if cleaned[0].isdigit():
        cleaned = f"u{cleaned}"[:24]
    return cleaned


def unique_username(email: str, name: str | None = None) -> str:
    base = _username_base(email, name)
    candidate = base
    n = 2
    while User.objects.filter(username__iexact=candidate).exists():
        suffix = str(n)
        candidate = f"{base[: 150 - len(suffix)]}{suffix}"
        n += 1
        if n > 10_000:  # pragma: no cover
            raise ValidationError({"username": "Could not allocate a username."})
    return candidate


def user_from_google(info: dict):
    """Find or create a USER from a verified Google payload. Returns (user, created)."""
    sub = str(info["sub"])
    email = str(info.get("email") or "").strip().lower()

    existing = User.objects.filter(google_sub=sub).first()
    if existing:
        if email and not existing.email:
            existing.email = email
            existing.save(update_fields=["email"])
        return existing, False

    if email:
        by_email = User.objects.filter(email__iexact=email).first()
        if by_email:
            if by_email.google_sub and by_email.google_sub != sub:
                raise GoogleAuthError("This Google account cannot be linked.")
            by_email.google_sub = sub
            update = ["google_sub"]
            if email and not by_email.email:
                by_email.email = email
                update.append("email")
            by_email.save(update_fields=update)
            mark_email_verified(by_email)
            return by_email, False

    user = User(
        username=unique_username(email, info.get("name")),
        email=email,
        role=Role.USER,
        google_sub=sub,
    )
    user.set_unusable_password()
    user.save()
    return user, True


class GoogleSignInUnavailable(APIException):
    status_code = 503
    default_detail = "Google sign-in is not configured."
    default_code = "google_unavailable"


def raise_as_api(exc: GoogleAuthError):
    """Map GoogleAuthError to a DRF exception."""
    if exc.status >= 500:
        err = GoogleSignInUnavailable()
        err.detail = exc.detail
        raise err
    raise ValidationError({"detail": exc.detail})
